flink-filesystems/flink-gs-fs-hadoop/src/main/java/org/apache/flink/fs/gs/writer/GSRecoverableWriterCommitter.java

```diff
@@ -0,0 +1,189 @@
+        List<GSBlobIdentifier> composeBlobIds = sourceBlobIdentifiers.subList(0, composeToIndex);
+        List<GSBlobIdentifier> remainingBlobIds =
+                sourceBlobIdentifiers.subList(composeToIndex, sourceBlobIdentifiers.size());
+
+        // determine the resulting blob id for this compose operation. if this is the last compose,
+        // i.e. if there are no remaining blob ids, then the composed blob id is the originally
+        // specified target blob id. otherwise, we must create an intermediate blob id to hold the
+        // result of this compose operation
+        GSBlobIdentifier composedBlobId =
+                remainingBlobIds.isEmpty()
+                        ? targetBlobIdentifier
+                        : BlobUtils.generateTemporaryBlobIdentifier(
+                                recoverable.finalBlobIdentifier, options);
+
+        // compose the blobs
+        storage.compose(composeBlobIds, composedBlobId);
+
+        // if we have remaining blobs, add the composed blob id to the beginning of the list
+        // of remaining blob ids, and recurse
+        if (!remainingBlobIds.isEmpty()) {
+            remainingBlobIds.add(0, composedBlobId);
+            composeBlobs(remainingBlobIds, targetBlobIdentifier);
+        }
+    }
+
+    /**
+     * Writes the final blob by composing the temporary blobs and copying, if necessary.
+     *
+     * @throws IOException On underlying failure.
+     */
+    private void writeFinalBlob() throws IOException {
+
+        // compose all the component blob ids into the final blob id. if the component blob ids are
+        // in the same bucket as the final blob id, this can be done directly. otherwise, we must
+        // compose to a new temporary blob id in the same bucket as the component blob ids and
+        // then copy that blob to the final blob location
+        String temporaryBucketName =
+                BlobUtils.getTemporaryBucketName(recoverable.finalBlobIdentifier, options);
+        if (recoverable.finalBlobIdentifier.bucketName.equals(temporaryBucketName)) {
+
+            // compose directly to final blob
+            composeBlobs(recoverable.getComponentBlobIds(options), recoverable.finalBlobIdentifier);
+
+        } else {
+
+            // compose to the intermediate blob, then copy
+            GSBlobIdentifier intermediateBlobIdentifier =
+                    BlobUtils.generateTemporaryBlobIdentifier(
+                            recoverable.finalBlobIdentifier, options);
+            composeBlobs(recoverable.getComponentBlobIds(options), intermediateBlobIdentifier);
+            storage.copy(intermediateBlobIdentifier, recoverable.finalBlobIdentifier);
+        }
+    }
+
+    /**
+     * Clean up after a successful commit operation, by deleting any temporary blobs associated with
+     * the final blob.
+     *
+     * @throws IOException On underlying storage failure
+     */
+    private void cleanupTemporaryBlobs() throws IOException {
```

Comment:
nit:
```suggestion
    private void cleanupTemporaryBlobs() {
```

Reply:
Done in [9e8472b](https://github.com/apache/flink/pull/15599/commits/9e8472bd88cbfa2cb7b606cf9c52f089aecabe6d).